transformer_engine/pytorch/module/base.py

```diff
@@ -386,7 +386,8 @@ def get_extra_state(self) -> torch.Tensor:
             # Store other pickelable values.
             extra = {}
             for k, v in self.fp8_meta.items():
-                if isinstance(v, (bool, int, float, str, tuple, list)):
+                if (k != 'buffer_index_and_autocast_key'
```

Comment:
Why is this check needed? I presume it's because we do a weird thing where `DotProductAttention` fills its state dict with the extra state from `FusedAttention`, but that just means the state dict is inconsistent with the model. Loading the state dict puts the extra state back in `FusedAttention`, so I'd expect `buffer_index_and_autocast_key` to still be valid.

If it's because there's a test failure at https://github.com/NVIDIA/TransformerEngine/blob/9a3e08f52109237e94d0da823b87ce2546562221/tests/pytorch/test_sanity.py#L1093 I'd prefer if we skipped `'buffer_index_and_autocast_key'` there rather than here.

Reply:
It's not to do with filling `DotProductAttention`'s state dict from `FusedAttention` or vice verse. It's because if `buffer_index_and_autocast_key` exists in the new state dict (of `DotProductAttention` or `FusedAttention`), the global amax buffers wouldn't be updated with the new buffers inside each module: https://github.com/NVIDIA/TransformerEngine/blob/905d94f487e8ee6c03203c79e94acea6396f6142/transformer_engine/pytorch/module/base.py#L326